hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/health/TestDeletingContainerHandler.java

```diff
@@ -0,0 +1,324 @@
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    Assert.assertFalse(deletingContainerHandler.handle(request));
+  }
+
+  @Test
+  public void testNonDeletingRatisContainerReturnsFalse() {
+    ContainerInfo containerInfo = ReplicationTestUtil.createContainerInfo(
+        ratisReplicationConfig, 1, CLOSED);
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicas(containerInfo.containerID(),
+            ContainerReplicaProto.State.CLOSING, 0, 0, 0);
+
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.EMPTY_LIST)
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    Assert.assertFalse(deletingContainerHandler.handle(request));
+  }
+
+  /**
+   * If a container is in Deleting state and no replica exists,
+   * change the state of the container to DELETED.
+   */
+  @Test
+  public void testCleanupIfNoReplicaExist()
+      throws IOException, TimeoutException, InvalidStateTransitionException {
+    //ratis container
+    cleanupIfNoReplicaExist(RatisReplicationConfig.getInstance(
+        HddsProtos.ReplicationFactor.THREE));
+
+    //ec container
+    cleanupIfNoReplicaExist(ecReplicationConfig);
+  }
+
+
+  private void cleanupIfNoReplicaExist(ReplicationConfig replicationConfig)
+      throws IOException, TimeoutException, InvalidStateTransitionException {
+    ContainerInfo containerInfo = containerManager.allocateContainer(
+        replicationConfig, "admin");
+    ContainerID cID = containerInfo.containerID();
+
+    //change the state of the container to Deleting
+    containerManager.updateContainerState(cID,
+        HddsProtos.LifeCycleEvent.FINALIZE);
+    containerManager.updateContainerState(cID, HddsProtos.LifeCycleEvent.CLOSE);
+    containerManager.updateContainerState(cID,
+        HddsProtos.LifeCycleEvent.DELETE);
+
+    Set<ContainerReplica> containerReplicas = new HashSet<>();
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.EMPTY_LIST)
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    Assert.assertTrue(deletingContainerHandler.handle(request));
+    Assert.assertTrue(containerInfo.getState() == DELETED);
+  }
+
+  /**
+   * If a container is in Deleting state , some replicas exist and
+   * for each replica there is a pending delete, then do nothing.
+   */
+  @Test
+  public void testNoNeedResendDeleteCommand()
+      throws IOException, TimeoutException, InvalidStateTransitionException {
+    //ratis container
+    ContainerInfo containerInfo = containerManager.allocateContainer(
```

Comment:
Rather than using CM to allocated a container, just create a ContainerInfo with the correct state - there are some helper test classes in `ReplicationTestUtil` we have been using for this.

Reply:
i used `ReplicationTestUtil` origianlly, but there will be a problem casued by `lifeCycleStateMap` when firing an event to change the state of a container, so i use containerManager#allocateContainer directly!
```
    //change the state of the container to Deleting
    containerManager.updateContainerState(cID,
        HddsProtos.LifeCycleEvent.FINALIZE);
    containerManager.updateContainerState(cID, HddsProtos.LifeCycleEvent.CLOSE);
    containerManager.updateContainerState(cID,
        HddsProtos.LifeCycleEvent.DELETE);
```

Reply:
To test this handler, we don't need to change the container state - we can just set it to the state we need when we created it and avoid the updateContainerState calls.

Reply:
here , we should verify that the container state is changed for DELETING to DELETED,  and we need this check to make sure the handler works as expected

Reply:
No we don't. These are unit tests and testing the transition of the container is another units job. We just need to test the correct method is called with the correct parameters on ReplicationManager and then from a unit testing perspective, we know the handler is doing the correct thing.

Reply:
ok, i have updated tests , please take a look